Don't destroy auto-save file written by different architecture...

Paul-Licameli opened the pull request:
... the error checking might not be complete, but it is sufficient for the
observed cases, where switching between 32 and 64 bit Mac builds causes
auto-recovery in one build to destroy the data saved by the other build.

Now instead, you will see the (previously existing) message
"Could not decode file: %s" which comes from AudacityProject::OpenFile.

# Pull Requests

If you are submitting a pull request, read https://wiki.audacityteam.org/wiki/GitHub_Pull_Requests 


## The key points: 

* Come over talk with us at the audacity devel email list. If you just rely on the GitHub pull request messages, you may find we ignore or close the pull request for what does not seem to you to be a good reason. Please come and talk. 

* Translators should subscribe to audacity translators email list instead.  The translators list is also the right place for most translation discussion. 

There is a bit more on our wiki about how we use the pull requests and the labels that can be attached to pull requests.

JamesCrook commented:
Is there any risk of this rejecting a normally created auto recovery file?  If I recall correctly, the handling of matching tags is done in a special way for recordings.  Might the tag checking in Lookup in some circumstance reject a file from a crash during recording?

JamesCrook commented:
Is there any risk of 'stuck with autosave' where a user who has a bad autosave file gets prompted with it every time they start Audacity?  (and no guidance what to do to recover from that)

Paul-Licameli commented:
Answering the first comment, no.

Where the tag lookup quietly failed previously, what happened was that the map mIds associated a number with the empty string, and the result would be a malformed empty XML tag "<>" written to the file -- causing exactly the original symptom of this bug.

If a recovery file is correct, this lookup failure will not happen.

Besides all of which, the recording recovery handling is quite separate from the saving and reinterpreting of the binary file format used by autosave.

Paul-Licameli commented:
Regarding the second question, when there is autosave file, the user is given the choice to recover (which may fail), discard the files, or quit the program.  If recovery fails, the directories are left in their prior state, so the dialog will appear again when you restart.

Discarding the recovery files would cause the dialog not to happen again at startup.

Since we are after string freeze, we can't improve the guidance.  "Could not decode file: <path>" is all you will see if recovery fails.

Maybe we could improve that now if we accept the non-translation of that message in other locales.

JamesCrook commented:
If this is a P2 bug, then an English-only (for 2.3.2) error message with better guidance is OK.  The error message will have a help button, and this will in any case take us to an English language help page.  The English message will get translated in 2.3.3.

Since the fix is intended to avoid dataloss, I think better guidance is called for, for the fix to actually avoid dataloss.  So if you want to continue with this, please provide an updated pull request.

Paul-Licameli commented:
Rebased and updated the error message.

Also can write the changed error message to standard out when recovery is invoked via command line.  That was the only other place in the program where the decoding of the binary autosave file happens.

JamesCrook commented:
Thanks.  Let's get this to QA and see how it fares.  Committed.
